src/problems/peg_game.clj

```diff
@@ -0,0 +1,144 @@
+;;             9 [[2 5] [7 8]]
+;;             10 [[3 6] [12 11]]
+;;             11 [[4 7] [13 12]]
+;;             12 [[3 7] [5 8] [10 11] [14 13]]
+;;             13 [[4 8] [11 12]]
+;;             14 [[5 9] [12 13]]))
+
+
+;;Diagonal increase left to right
+;;[(+ (+ vector-index row-number) (+ row-number 1)) (+ vector-index row-number)]
+
+;;Diagonal increase right to left
+;;[(+ (+ vector-index (+ row-number 1)) (+ row-number 2)) (+ vector-index (+ row-number 1))]
+
+;;Diagonal decrease left to right
+;;[(- (- vector-index row-number) (- row-number 1)) (- vector-index row-number)]
+
+;;Diagonal decrease right to left
+;;[(- (- vector-index (- row-number 1)) (- row-number 2)) (- vector-index (- row-number 1))]
+
+(defn num-rows [board]
+  (loop [index 1
+         total-index (count board)
+         current-index 0
+         number-of-rows 0]
+    (if (not= total-index current-index)
+      (recur (+ index 1) total-index (+ current-index index) (+ number-of-rows 1))
+      number-of-rows)))
+
+
+(defn generate-moves [board vector-index row-index row-number number-of-rows hashmap]
+  [
+    (if (> row-index 2) [(- vector-index 2) (- vector-index 1)])
+    (if (< row-index (- row-number 1)) [(+ vector-index 2) (+ vector-index 1)])
+    (if (> row-index 2) [(- (- vector-index row-number) (- row-number 1)) (- vector-index row-number)])
+    (if (< row-index (- row-number 1)) [(- (- vector-index (- row-number 1)) (- row-number 2)) (- vector-index (- row-number 1))])
+    (if (< row-number (- number-of-rows 1)) [(+ (+ vector-index (+ row-number 1)) (+ row-number 2)) (+ vector-index (+ row-number 1))])
+    (if (< row-number (- number-of-rows 1)) [(+ (+ vector-index row-number) (+ row-number 1)) (+ vector-index row-number)])
+  ])
+
+
+
+
+(defn make-moves
+  "makes the moves that are possible for the board given to it"
+  [board]
+  (loop [board board
+         vector-index 0
+         row-index 1
+         row-number 1
+         number-of-rows (num-rows board)
+         hashmap {}]
+    (if (not= vector-index (count board))
+      (if (not= row-index row-number)
+        (recur board (+ vector-index 1) (+ row-index 1) row-number number-of-rows (assoc hashmap vector-index (generate-moves board vector-index row-index row-number number-of-rows hashmap)))
+        (recur board (+ vector-index 1) 1 (+ row-number 1) number-of-rows (assoc hashmap vector-index (generate-moves board vector-index row-index row-number number-of-rows hashmap))))
+      hashmap)))
+
+
+
+;;Need to put the same size or larger peg game as the one you are actually performing the search on.
+
+(def moves
+ (make-moves [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]))
+
+
+;; Heuristic perfers empty spaces on the bottom of the board.
+
+(defn heuristic
+  "Heuristic perfers empty spaces on the bottom of the board."
+  [parent child]
+  (let [heuristic (get (frequencies (subvec child (quot (count child) 2) (count child))) 0)]
+    (if (= heuristic nil)
+      0
+      (- 0 heuristic))))
```

Comment:
This is potentially problematic since you can return a negative number. This can in fact "cancel out" the cost of steps in A* in weird ways, since a negative value for the heuristic is essentially saying it would take a *negative* number of steps/work/effort to solve the problem from this state, which would be a pretty nifty trick. 😄 

It won't really hurt in this puzzle because all the steps are equal cost and there's no backtracking, but it could be a problem in other settings. I would have instead counted the number of 1's in the bottom half of the board, and just returned that. Then you'd be minimizing a value that would eventually drop to zero when you had all blanks on the bottom.

All that said, why does this heuristic make sense? It seems to work (and quite well), but it's not obvious to me why the bottom of the board is a good place to focus on clearing first.